jdk12-reference/src/main/java/org/jstlang/util/ObjectUnflattener.java

```diff
@@ -0,0 +1,104 @@
+package org.jstlang.util;
+
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import lombok.NonNull;
+import lombok.RequiredArgsConstructor;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@RequiredArgsConstructor(staticName = "getDefault")
+public class ObjectUnflattener implements Function<Object, Object> {
+
+    private final @NonNull String splitEscape = "\\";
+    private final @NonNull String splitterExpression = "(?<!\\\\)\\.";
+
+    /**
+     * This is the entry point to the Object Unflattener logic.
+     * @param input - Any Object
+     * @return - An unflattened version of the Object
+     */
+    public Object apply(Object input) {
+        return this.typeChecker(input);
+    }
+
+    /**
+     * This is the entry point to the Object Unflattener logic.
+     * This method takes a Map and will check if every key in the map can be unflattened based on the splitting criteria.
+     * This action is then performed recursively over every value in the Map, leading to a Depth First traversal and
+     * transformation of the initial input structure
+     * @param input
+     * @return
+     */
+    public Map<String,Object> unflattenMap(Map<String,Object> input) {
+        Map<String,Object> unflattenedMap = Maps.newHashMap();
```

Comment:
LGTM I prefer the default behavior to stay with insertion order since the input could be in its insertion order which is the read-order.